src/core/cliManager.ts

```diff
@@ -97,143 +103,342 @@ export class CliManager {
-					"Downloaded binary size is",
-					prettyBytes(newStat?.size || 0),
+					"Moving existing binary to",
+					path.basename(oldBinPath),
 				);
+				await fs.rename(binPath, oldBinPath);
+			}
 
-				// Make sure we can execute this new binary.
-				const version = await cliUtils.version(binPath);
-				this.output.info("Downloaded binary version is", version);
+			// Step 2: Move temp to final location
+			this.output.info("Moving downloaded file to", path.basename(binPath));
+			await fs.rename(tempFile, binPath);
+		} catch (error) {
+			throw cliUtils.maybeWrapFileLockError(error, binPath);
+		}
+
+		// For debugging, to see if the binary only partially downloaded.
+		const newStat = await cliUtils.stat(binPath);
+		this.output.info(
+			"Downloaded binary size is",
+			prettyBytes(newStat?.size || 0),
+		);
+
+		// Make sure we can execute this new binary.
+		const version = await cliUtils.version(binPath);
+		this.output.info("Downloaded binary version is", version);
+	}
 
+	/**
+	 * Unified handler for any binary-related failure.
+	 * Checks for existing or old binaries and prompts user once.
+	 */
+	private async handleAnyBinaryFailure(
+		error: unknown,
+		binPath: string,
+		expectedVersion: string,
+	): Promise<string> {
+		const message =
+			error instanceof cliUtils.FileLockError
+				? "Unable to update the Coder CLI binary because it's in use"
+				: "Failed to update CLI binary";
+
+		// Try existing binary first
+		const existingCheck = await this.checkBinaryVersion(
+			binPath,
+			expectedVersion,
+		);
+		if (existingCheck.version) {
+			// Perfect match - use without prompting
+			if (existingCheck.matches) {
 				return binPath;
 			}
-			case 304: {
-				this.output.info("Using existing binary since server returned a 304");
+			// Version mismatch - prompt user
+			if (await this.promptUseExistingBinary(existingCheck.version, message)) {
 				return binPath;
 			}
-			case 404: {
-				vscode.window
-					.showErrorMessage(
-						"Coder isn't supported for your platform. Please open an issue, we'd love to support it!",
-						"Open an Issue",
-					)
-					.then((value) => {
-						if (!value) {
-							return;
-						}
-						const os = cliUtils.goos();
-						const arch = cliUtils.goarch();
-						const params = new URLSearchParams({
-							title: `Support the \`${os}-${arch}\` platform`,
-							body: `I'd like to use the \`${os}-${arch}\` architecture with the VS Code extension.`,
-						});
-						const uri = vscode.Uri.parse(
-							`https://github.com/coder/vscode-coder/issues/new?${params.toString()}`,
-						);
-						vscode.env.openExternal(uri);
-					});
-				throw new Error("Platform not supported");
+		}
+
+		// Try .old-* binaries as fallback
+		const oldBinaries = await cliUtils.findOldBinaries(binPath);
+		if (oldBinaries.length > 0) {
+			const oldCheck = await this.checkBinaryVersion(
+				oldBinaries[0],
+				expectedVersion,
+			);
+			if (
+				oldCheck.version &&
+				(oldCheck.matches ||
```

Comment:
Why do we only try _one_ old binary when we are tracking all of them?

I'm trying to think about _when_ this condition actually might happen. I.e. we have an old binary _and_ it's the right version. This means you either downgraded coderd or switched deployment. In either case, _any_ of the old binaries may be the correct one.